### PR TITLE
Fix integer overflow for Gregorian calendar by using saturating arithmetic ops

### PR DESCRIPTION
**Pull request by sotam1069**
Decide on overflow behavior for calendar arithmetic https://github.com/unicode-org/icu4x/issues/2151

### CODE REVIEW

File: components/calendar/src/gregorian.rs [hunk 1]
```diff
@@ -163,7 +163,7 @@ impl Date<Gregorian> {
     ///
     /// Years are specified as ISO years.
     ///
-    /// ```rust
+    /// ```rust
```
Comment (echeran) [inline components/calendar/src/gregorian.rs:166]: Run `cargo fmt` to make the CI job pass. It likes to remove trailing whitespace at the end of lines.

File: components/calendar/src/gregorian.rs [hunk 2]
```diff
@@ -228,8 +228,26 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
-            number: 1 - year,
+            number: 1_i32.saturating_sub(year),
             related_iso: None,
         }
     }
 }
+
+
+#[cfg(test)]
+
```
Comment (echeran) [inline components/calendar/src/gregorian.rs:239]: Remove blank line between `#[...]` line and the following line that it applies to.

File: components/calendar/src/gregorian.rs [hunk 3]
```diff
@@ -228,8 +228,26 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
+mod test {
+    use super::*;
+
+    #[test]
+    fn day_of_year_info_overflow(){
+        let date_max = Date::try_new_gregorian_date(i32::MAX, 4, 4).unwrap();
+        //Compares the next_year variable to i32::MAX to check for overflow
```
Comment (echeran) [inline components/calendar/src/gregorian.rs:246]:
```suggestion
        // Compares the next_year variable to i32::MAX to check for overflow
```

File: components/calendar/src/gregorian.rs [hunk 4]
```diff
@@ -228,8 +228,26 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
+        assert_eq!(Calendar::day_of_year_info(&Gregorian, &date_max.inner).next_year.number,i32::MAX,);
+        
+        //Compares the prev_year variable to i32::MIN to check for overflow
```
Comment (echeran) [inline components/calendar/src/gregorian.rs:249]:
```suggestion
        // Compares the prev_year variable to i32::MIN to check for overflow
```

File: components/calendar/src/gregorian.rs [hunk 5]
```diff
@@ -228,8 +228,26 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
 }
+
+
+#[cfg(test)]
+
+mod test {
+    use super::*;
+
+    #[test]
+    fn day_of_year_info_overflow(){
```
Comment (echeran) [inline components/calendar/src/gregorian.rs:244]: Space between the `()` and the `{`. The formatter will probably make that change.

File: components/calendar/src/gregorian.rs [hunk 6]
```diff
@@ -228,8 +228,26 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
+        let date_max = Date::try_new_gregorian_date(i32::MAX, 4, 4).unwrap();
+        //Compares the next_year variable to i32::MAX to check for overflow
+        assert_eq!(Calendar::day_of_year_info(&Gregorian, &date_max.inner).next_year.number,i32::MAX,);
```
Comment (echeran) [inline components/calendar/src/gregorian.rs:247]: (formatting again) Trailing comma inside parens.

File: components/calendar/src/iso.rs [hunk 7]
```diff
@@ -202,8 +203,8 @@ impl Calendar for Iso {
     }
 
     fn day_of_year_info(&self, date: &Self::DateInner) -> types::DayOfYearInfo {
```
Comment (sffc) [inline components/calendar/src/iso.rs:205]: Issue: don't have iso.rs changes in this PR

**Review by sffc — CHANGES_REQUESTED** [review]

**Review by echeran — DISMISSED** [review]
LGTM

File: components/calendar/src/gregorian.rs [hunk 8]
```diff
@@ -228,8 +228,109 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
+        ];
+
+        for case in cases {
+            let date = Date::try_new_gregorian_date(case.year, case.month, case.day).unwrap();
+
+            assert_eq!(
+                Calendar::day_of_year_info(&Gregorian, &date.inner)
+                    .prev_year
+                    .number,
+                case.prev_year.saturating_neg(),
```
Comment (sffc) [inline components/calendar/src/gregorian.rs:331]: Nit: If you always call `.saturating_neg()` on `case.prev_year`, then you should just represent `prev_year` in that form in the test case (as a positive year number).

File: components/calendar/src/gregorian.rs [hunk 9]
```diff
@@ -228,8 +228,109 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
...
+                year: i32::MIN,
+                month: 2,
+                day: 2,
+                prev_year: i32::MIN,
+            },
+            MinCase {
+                year: i32::MIN + 1,
+                month: 1,
+                day: 1,
+                prev_year: i32::MIN,
```
Comment (sffc) [inline components/calendar/src/gregorian.rs:320]: Nit: Add at least one cases that does not saturate

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/calendar/src/gregorian.rs [hunk 10]
```diff
@@ -228,8 +228,127 @@ pub(crate) fn year_as_gregorian(year: i32) -> types::FormattableYear {
     } else {
         types::FormattableYear {
             era: types::Era(tinystr!(16, "bce")),
-            number: 1 - year,
+            number: year.saturating_sub(0),
```
Comment (sffc) [inline components/calendar/src/gregorian.rs:231]: Issue: This is definitely different behavior. This is not correct. The year is being projected into BCE so it should be 1 - year. Please change back to what you had before, and fix the tests.

**Review by sffc — CHANGES_REQUESTED** [review]

**Review by sffc — APPROVED** [review]

**Review by Manishearth — COMMENTED** [review]
post-merge review: lgtm!

**Pull request merged**